Develop - Merge ArtifactSummary & DirectorySource from feature/indexed-source

wmrutten opened the pull request:
Not merged: BlobDb, FhirArchive

ewoutkramer commented on src/Hl7.Fhir.Specification/Specification/Source/Summary/ArtifactSummaryDetails.cs:
```diff
@@ -0,0 +1,200 @@
+
+    /// <summary>For accessing common artifact summary details.</summary>
+    /// <remarks>Common details are initialized by the <see cref="ArtifactSummaryGenerator"/> itself.</remarks>
+    public static class ArtifactSummaryDetails
```
Doesn't this mean that all extension methods on IArtifactSummaryDetailsProvider are in scope all the time?  So, now there's no real notion of "these attributes are right for a ValueSet"?

ewoutkramer replied:
Since this are all extension methods on static classes, why do we still have an ArtifactSummaryFactory? There are no specific classes of instances to create?

wmrutten replied:
1) Correct, I removed all concrete subclasses to simplify the default API design. As a consequence, we lose some type safety.
2) Client can override factor to create specific subclasses with additional type-safe properties. However by default, the API generates standard ArtifactSummary instances.

ewoutkramer replied:
Since we've chosen to have no subclasses, I think we should ditch the factory pattern for ArtifactSummaries then.  Your namespace will contain those extension methods, so for the API user this will only create confusion.

ewoutkramer commented on src/Hl7.Fhir.Specification/Specification/Source/DirectorySourceSettings.cs:
```diff
@@ -0,0 +1,273 @@
+        /// <see cref="NavigatorStreamFactory"/> delegate. This allows clients to implement
+        /// support for alternative serialization formats.
+        /// </remarks>
+        public NavigatorStreamFactory StreamFactory { get; set; } // = DefaultNavigatorStreamFactory.Create;
```
To align with the rest of the API, I think I prefer an event-based approach here, instead of a factory.

Settings.OnNavigatorStreamNeeded, with the given path as a parameter to the event

wmrutten replied:
Sure! And also rewrite SummaryFactory delegate to event?

ewoutkramer replied:
I now understand you want to do this, not so much to support customization of creation of the navigator stream, but to have an extensible way to support more (future) formats of FHIR.  This is a good idea, but we should generalize that - i.e. how would I know which parser to create for a given format?  (not just the stream).  Let's think about that for a bit before deciding on this.

ewoutkramer replied:
I'd like to know what @christiaan and @martijn think about this too.

ewoutkramer commented on src/Hl7.Fhir.Specification/Specification/Source/Summary/ArtifactSummaryDetailsCollection.cs:
```diff
@@ -0,0 +1,54 @@
+    /// Implemented by <see cref="ArtifactSummaryDetailsCollection"/> as well as <see cref="ArtifactSummary"/>.
+    /// Allows common extension methods for retrieving specific summary details.
+    /// </remarks>
+    public interface IArtifactSummaryDetailsProvider
```
Can we not combine ArtifactSummary and ArtifactSummaryCollection?  Why would we keep both?

wmrutten replied:
I decided to encapsulate the backing store. Client code just needs an indexer property this[key], so this is the interface expresses. Implementation is free to choose a specific back store, we could change/optimize this at a later time.
However IReadOnlyDictionary seems to fit this use case, so we could decide to use this and drop the custom interface. At the cost of a bit more work when we'd want to change the backing store (to implement the additional methods).

I want to prevent consumers from changing harvested properties. However I also don't want to repeat all this property logic. So I decided on a kind of builder pattern. The extractor fills a read/write collection, then hands it over to initialize a read-only summary instance, which saves and wraps the r/w collection. By implementing a common interface, we can re-use extension methods on both the r/w collection class and the final r/o summary instance.

ewoutkramer commented on src/Hl7.Fhir.Specification/Specification/Source/Summary/ArtifactSummaryDetails.cs:
```diff
@@ -0,0 +1,200 @@
+                // Explicit extractor chaining
+                if (ConformanceSummaryDetails.Extract(nav, details))
+                {
+                    nav.TryExtractValue(details, StructureDefinitionKindKey, "kind");
```
The boolean results from TryExtractValue() are not processed into the return value of the extraction?

wmrutten replied:
Correct, usually the caller can (and will) safely ignore the value. In some circumstances, it may be useful for callers to detect if a match was found.

ewoutkramer commented on src/Hl7.Fhir.Specification/Specification/Source/Summary/ArtifactSummaryGenerator.cs:
```diff
@@ -0,0 +1,203 @@
+                        var props = new ArtifactSummaryDetailsCollection();
+
+                        // Add default summary details
+                        props[ArtifactSummaryDetails.OriginKey] = origin;
```
Waarom zijn deze regels niet in een extractor gestopt?

ewoutkramer replied:
Because all that information does not come from an input navigator ;-)

wmrutten commented on src/Hl7.Fhir.Specification.Tests/Snapshot/SnapshotGeneratorTest.cs:
```diff
@@ -4580,32 +4580,27 @@ public void TestTypeProfileWithChildElementBinding()
             Base = ModelInfo.CanonicalUriForFhirCoreType(FHIRDefinedType.QuestionnaireResponse),
```
Nope, this shouldn't be part of this commit, not related to DirectorySource at all.
Actually, this unit test is commented out on DSTU2 branch.

wmrutten commented on src/Hl7.Fhir.Specification/Specification/Source/IConformanceSource.cs:
```diff
@@ -24,6 +22,9 @@ public interface IConformanceSource : IResourceResolver
         /// <returns></returns>
         IEnumerable<string> ListResourceUris(ResourceType? filter = null);
 
+        // [WMR 20171011] New
```
Yes, I think IConformanceSource should also include the following new method:
IEnumerable<ArtifactSummary> List()

wmrutten commented on src/Hl7.Fhir.Specification/Specification/Source/Summary/ArtifactSummaryGenerator.cs:
```diff
@@ -0,0 +1,203 @@
+    /// <param name="details">A collection of summary details extracted from the artifact.</param>
+    /// <param name="error">A error that occured while extracting details from the artifact, or <c>null</c>.</param>
+    /// <returns>A new <see cref="ArtifactSummary"/> instance.</returns>
+    public delegate ArtifactSummary ArtifactSummaryFactory(string typeName, ArtifactSummaryDetailsCollection details, Exception error = null);
```
As explained above, the extractor(s) first initialize a r/w collection. Then the extractor calls factory method to create a r/o Summary instance, initialized from the r/w collection.

wmrutten commented on src/Hl7.Fhir.Specification/Specification/Source/Summary/ArtifactSummaryNavigationExtensions.cs:
```diff
@@ -0,0 +1,117 @@
+        /// <param name="nav">An <see cref="IElementNavigator"/> instance.</param>
+        /// <param name="element">An element name.</param>
+        /// <returns><c>true</c> if, upon return, the navigator is positioned on a matching element, or <c>false</c> otherwise.</returns>
+        public static bool Find(this IElementNavigator nav, string element)
```
Sure, but we have to make sure the semantics fit the generic case.
These extension methods were specifically designed to facilitate the extraction process.

wmrutten commented on src/Hl7.Fhir.Specification/Specification/Source/Summary/ArtifactSummaryNavigationExtensions.cs:
```diff
@@ -0,0 +1,117 @@
+        /// <param name="key">A collection key.</param>
+        /// <param name="element">An element name.</param>
+        /// <param name="childElement">A child element name.</param>
+        public static bool TryExtractValue(this IElementNavigator nav, ArtifactSummaryDetailsCollection details, string key, string element, string childElement)
```
I don't think so....
All methods are expected to advance to the matching element, if found.
The child element value is extracted using TryExtractValue, which DOES clone the navigator, so the original navigator remains positioned on the matching element.
Behavior is tested and confirmed by unit tests (but we might need some more...).

ewoutkramer replied:
So, maybe the documentation should be clearer that these are specific "consumer" or "munch" methods, that take a very specific set of parameters

wmrutten commented on src/Hl7.Fhir.Specification/Specification/Source/Summary/ArtifactSummaryNavigationExtensions.cs:
```diff
@@ -0,0 +1,117 @@
+            var value = nav.Value; // ?.ToString();
+            if (value != null)
+            {
+                // Assumption: navigator always returns string values (?)
```
Will do!

wmrutten commented on src/Hl7.Fhir.Specification/Specification/Source/Summary/ArtifactSummaryNavigationExtensions.cs:
```diff
@@ -0,0 +1,117 @@
+        /// <param name="key">A collection key.</param>
+        /// <param name="element">An element name.</param>
+        /// <param name="childElement">A child element name.</param>
+        public static bool TryExtractValues(this IElementNavigator nav, ArtifactSummaryDetailsCollection details, string key, string element, string childElement)
```
Hmmm... thx ;p

wmrutten commented:
Pushed an updated version for review.
Tried to simplify and improve the design, in response to code review.
Renamed some types for clarification.
Added unit test to demonstrate/test custom harvester implementation.

wmrutten commented:
Updated ArtifactSummary design in response to code review.
Merged updated logic from feature/indexed-source into develop (DSTU2).
Does not include new experimental NonReentrantResolver (moved to separate feature branch).
